1_fetch/src/fetch_wqp_data.R

```diff
@@ -47,12 +65,62 @@ add_download_groups <- function(sitecounts_df, max_sites = 500, max_results = 25
                                               maxgroupsize = max_sites),
                download_grp = paste0(grid_id,"_",task_num))
     }) %>%
-    mutate(site_n = row_number()) %>%
-    select(site_id, results_count, site_n, download_grp)
+    mutate(pull_by_id = TRUE)
   
-  return(sitecounts_grouped)
+  # Assign a separate task number and download group for all sites with 
+  # bad identifiers within a grid
+  sitecounts_grouped_bad_ids <- sitecounts_bad_ids %>%
+    group_by(grid_id) %>%
+    mutate(task_num = max(sitecounts_grouped_good_ids$task_num) + cur_group_id(),
+           download_grp = paste0(grid_id,"_",task_num),
+           pull_by_id = FALSE) %>%
+    ungroup()
+  
+  # Combine all sites back together, now with assigned download_grp id's
+  # and format columns
+  sitecounts_grouped_out <- sitecounts_grouped_good_ids %>%
+    bind_rows(sitecounts_grouped_bad_ids) %>%
+    arrange(download_grp) %>%
+    mutate(site_n = row_number()) %>% 
+    select(site_id, lat, lon, datum, results_count, site_n, download_grp, pull_by_id)
+  
+  return(sitecounts_grouped_out)
+
+}
+
 
+
+#' Function to return a small bounding box around site(s)
+#' 
+#' @details Some site identifiers contain undesired characters and cannot be
+#' parsed by WQP. This function creates a small bounding box around the 
+#' site(s) for the purposes of querying WQP by bBox rather than site id. 
+#' 
+#' @param sites data frame containing at minimum columns lat, lon, and datum
+#' @param buffer_dist_degrees double; value indicating how large of a buffer
+#' should be added to the site(s). Bounding box will be computed in WGS84 and
+#' units are in degrees. Defaults to 0.005 degrees (~500 m).
+#' 
+create_site_bbox <- function(sites, buffer_dist_degrees = 0.005){
```

Comment:
I'll note that this use of ~500 m as a buffer distance for creating a small bounding box is arbitrary. I've included that distance as an argument in `create_site_bbox()` but I currently don't pass that flexibility along as an argument to `fetch_wqp_data()` because I wanted to hide that complexity from the user. 

~500 m works for all of the edge cases in this example, but I wouldn't be too surprised if it fails. The edge cases in this example are all lakes, and I noticed that for some, 100 m was enough of a buffer, but for others, 500 m is needed in order to get data back from the WQP query. I suspect this has to do with lake size, but I'm not sure. Let me know if you have any other ideas on this tradeoff between robustness and complexity.